zap/src/output/tooling.rs

```diff
@@ -0,0 +1,347 @@
+		self.indent();
+
+		self.push_line(&format!("Name = \"{}\",", ev.name));
+		self.push_line("Arguments = { value }");
+
+		self.dedent();
+		self.push_line("})");
+
+		self.dedent();
+	}
+
+	fn push_function_callback(&mut self, first: bool, fn_decl: &FnDecl) {
+		let id = fn_decl.id;
+
+		self.push_indent();
+
+		if first {
+			self.push("if ");
+		} else {
+			self.push("elseif ");
+		}
+
+		// push_line is not used here as indent was pushed above
+		// and we don't want to push it twice, especially after
+		// the if/elseif
+		self.push(&format!("id == {id} then"));
+		self.push("\n");
+
+		self.indent();
+
+		self.push_line("if isServer then");
+		self.indent();
+
+		self.push_line("local value");
+
+		if let Some(data) = &fn_decl.rets {
+			self.push_stmts(&des::gen(data, "value", true));
+		}
+
+		self.push_line("table.insert(events, {");
+		self.indent();
+
+		self.push_line(&format!("Name = \"{} (callback)\",", fn_decl.name));
+		self.push_line("Arguments = { value }");
+
+		self.dedent();
+		self.push_line("})");
+		
+		self.dedent();
+		self.push_line("else");
+		self.indent();
+
+		self.push_line("local value");
+
+		if let Some(data) = &fn_decl.args {
+			self.push_stmts(&des::gen(data, "value", true));
+		}
+
+		self.push_line("table.insert(events, {");
+		self.indent();
+
+		self.push_line(&format!("Name = \"{} (request)\",", fn_decl.name));
+		self.push_line("Arguments = { value }");
+
+		self.dedent();
+		self.push_line("})");
+
+		self.dedent();
+		self.push_line("end");
+
+
+		self.dedent();
+	}
+
+	pub fn output(mut self) -> String {
+		self.push_line("--!native");
+		self.push_line("--!optimize 2");
+		self.push_line("--!nocheck");
+		self.push_line("--!nolint");
+		self.push_line("--#selene: allow(incorrect_standard_library_use)");
+
+		self.push_line(&format!(
+			"-- Tooling generated by Zap v{} (https://github.com/red-blox/zap)",
+			env!("CARGO_PKG_VERSION")
+		));
+
+		// if self.config.evdecls.is_empty() && self.config.fndecls.is_empty() {
+		// 	return self.buf;
+		// };
+
+		self.push_line("return function(player, incoming_buff, incoming_inst)");
+		self.indent();
+
+		self.push_line("local isServer = true");
+		self.push_line("if type(player) == \"buffer\" then");
+		self.indent();
+
+		self.push_line("isServer = false");
+		self.push_line("incoming_inst = incoming_buff");
+		self.push_line("incoming_buff = player");
+		self.push_line("player = nil");
+
+		self.dedent();
+		self.push_line("end");
+		self.push("\n");
+
+		self.push_line("local incoming_read = 0");
+		self.push_line("local incoming_ipos = 0");
+
+		self.push_line("local function read(len: number)");
+		self.indent();
+
+		self.push_line("local pos = incoming_read");
+		self.push_line("incoming_read = incoming_read + len");
+		self.push("\n");
+		self.push_line("return pos");
+
+		self.dedent();
+		self.push_line("end");
+		self.push_line("local len = buffer.len(incoming_buff)");
+		self.push("\n");
+
+		self.push_tydecls();
```

Comment:
One of the major intents was not to touch the rest of the codebase (e.g `irgen`), and that meant that the types functions have to be inside the function. Let me know if you think this is an issue.

Reply:
![RobloxStudio-njRe63Fu@2x](https://github.com/user-attachments/assets/733b122f-3948-47c5-98b3-b031190478e6)

This doesn't seem to be as slow as I was expecting it to be...

Reply:
@jackdotink - before I merge can you confirm if this is ok?

Reply:
lgtm